File: dart_sample/lib/demos/material/color_palette_demo_page.dart
import 'package:material_ui/material_ui.dart';

class ColorPaletteDemoPage extends StatefulWidget {
  const ColorPaletteDemoPage({super.key});

  @override
  State<ColorPaletteDemoPage> createState() => _ColorPaletteDemoPageState();
}

class _ColorPaletteDemoPageState extends State<ColorPaletteDemoPage> {
  static const List<(String, MaterialColor)> _swatches = <(String, MaterialColor)>[
    ('blue', Colors.blue),
    ('green', Colors.green),
    ('deepOrange', Colors.deepOrange),
    ('grey', Colors.grey),
  ];

  static const List<int> _shades = <int>[
    50,
    100,
    200,
    300,
    400,
    500,
    600,
    700,
    800,
    900,
  ];

  int _swatchIndex = 0;
  bool _useMaterial3 = false;

  @override
  Widget build(BuildContext context) {
    final (String name, MaterialColor swatch) = _swatches[_swatchIndex];
    final ThemeData pageTheme = ThemeData(
      useMaterial3: _useMaterial3,
      primarySwatch: swatch,
    );

    return Theme(
      data: pageTheme,
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: <Widget>[
          const Text('Colors + primarySwatch', style: TextStyle(fontSize: 20)),
          const SizedBox(height: 8),
          const Text(
            'MaterialColor shades, ColorScheme.fromSwatch and the '
            'primarySwatch-derived theme colors.',
            style: TextStyle(fontSize: 14, color: Color(0x8A000000)),
          ),
          const SizedBox(height: 10),
          Row(
            children: <Widget>[
              _buildControlButton(
                label: 'swatch: $name',
                onTap: () => setState(
                  () => _swatchIndex = (_swatchIndex + 1) % _swatches.length,
                ),
                width: 150,
                background: const Color(0xFFE9F0FF),
              ),
              const SizedBox(width: 8),
              _buildControlButton(
                label: _useMaterial3 ? 'M3' : 'M2',
                onTap: () => setState(() => _useMaterial3 = !_useMaterial3),
                width: 80,
                background: const Color(0xFFEAF6F7),
              ),
            ],
          ),
          const SizedBox(height: 10),
          Expanded(
            child: SingleChildScrollView(
              child: Column(
                children: <Widget>[
                  _buildShadeStrip(name, swatch),
                  const SizedBox(height: 14),
                  _buildThemeProbe(pageTheme),
                ],
              ),
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildShadeStrip(String name, MaterialColor swatch) {
    return Container(
      color: const Color(0xFFF7F9FC),
      padding: const EdgeInsets.all(12),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: <Widget>[
          Text(
            'Colors.$name shades',
            style: const TextStyle(fontSize: 14, color: Colors.black),
          ),
          const SizedBox(height: 8),
          Row(
            children: <Widget>[
              for (final int shade in _shades)
                Expanded(
                  child: Container(
                    height: 56,
                    color: swatch[shade],
                    alignment: Alignment.center,
                    child: Text(
                      '$shade',
                      style: TextStyle(
                        fontSize: 11,
                        color:
                            ThemeData.estimateBrightnessForColor(
                                  swatch[shade]!,
                                ) ==
                                Brightness.dark
                            ? Colors.white
                            : Colors.black,
                      ),
                    ),
                  ),
                ),
            ],
          ),
        ],
      ),
    );
  }

  Widget _buildThemeProbe(ThemeData theme) {
    return Container(
      color: const Color(0xFFF7F9FC),
      padding: const EdgeInsets.all(12),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: <Widget>[
          const Text(
            'Theme colors',
            style: TextStyle(fontSize: 14, color: Colors.black),
          ),
          const SizedBox(height: 8),
          _buildSwatchRow('colorScheme.primary', theme.colorScheme.primary),
          _buildSwatchRow('colorScheme.secondary', theme.colorScheme.secondary),
          _buildSwatchRow('primaryColor', theme.primaryColor),
          _buildSwatchRow('primaryColorLight', theme.primaryColorLight),
          _buildSwatchRow('primaryColorDark', theme.primaryColorDark),
          _buildSwatchRow('canvasColor', theme.canvasColor),
          _buildSwatchRow('cardColor', theme.cardColor),
        ],
      ),
    );
  }

  Widget _buildSwatchRow(String label, Color color) {
    return Row(
      children: <Widget>[
        Container(width: 28, height: 20, color: color),
        const SizedBox(width: 10),
        Expanded(
          child: Text(
            label,
            style: const TextStyle(fontSize: 12, color: Colors.black),
          ),
        ),
        Text(
          '#${color.value.toRadixString(16).toUpperCase().padLeft(8, '0')}',
          style: const TextStyle(fontSize: 12, color: Color(0xFF607D8B)),
        ),
      ],
    );
  }

  Widget _buildControlButton({
    required String label,
    required VoidCallback onTap,
    required double width,
    required Color background,
  }) {
    return SizedBox(
      width: width,
      child: TextButton(
        onPressed: onTap,
        style: TextButton.styleFrom(
          backgroundColor: background,
          foregroundColor: Colors.black,
          minimumSize: const Size(0, 36),
          padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 8),
          shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(8)),
        ),
        child: Text(label, style: const TextStyle(fontSize: 12)),
      ),
    );
  }
}

File: src/Sample/Plumix.Sample/Demos/Material/ColorPaletteDemoPage.cs
using System;
using System.Collections.Generic;
using Avalonia;
using Avalonia.Media;
using Plumix.Material;
using Plumix.Rendering;
using Plumix.Widgets;

// Dart parity source (reference): dart_sample/lib/demos/material/color_palette_demo_page.dart (exact sample parity)

namespace Plumix;

public sealed class ColorPaletteDemoPage : StatefulWidget
{
    public override State CreateState()
    {
        return new ColorPaletteDemoPageState();
    }
}

internal sealed class ColorPaletteDemoPageState : State
{
    private static readonly IReadOnlyList<(string Name, MaterialColor Swatch)> Swatches =
    [
        ("blue", MaterialColors.Blue),
        ("green", MaterialColors.Green),
        ("deepOrange", MaterialColors.DeepOrange),
        ("grey", MaterialColors.Grey),
    ];

    private static readonly int[] Shades = [50, 100, 200, 300, 400, 500, 600, 700, 800, 900];

    private int _swatchIndex;
    private bool _useMaterial3;

    public override Widget Build(BuildContext context)
    {
        (string name, MaterialColor swatch) = Swatches[_swatchIndex];
        var pageTheme = new ThemeData(
            useMaterial3: _useMaterial3,
            primarySwatch: swatch);

        return new Theme(
            data: pageTheme,
            child: new Column(
                crossAxisAlignment: CrossAxisAlignment.Stretch,
                spacing: 10,
                children:
                [
                    new Text("Colors + primarySwatch", fontSize: 20, color: Colors.Black),
                    new Text(
                        "MaterialColor shades, ColorScheme.fromSwatch and the primarySwatch-derived theme colors.",
                        fontSize: 14,
                        color: Color.Parse("#8A000000")),
                    new Row(
                        spacing: 8,
                        children:
                        [
                            BuildControlButton(
                                label: $"swatch: {name}",
                                onTap: () => SetState(() =>
                                    _swatchIndex = (_swatchIndex + 1) % Swatches.Count),
                                width: 150,
                                background: Color.Parse("#FFE9F0FF")),
                            BuildControlButton(
                                label: _useMaterial3 ? "M3" : "M2",
                                onTap: () => SetState(() => _useMaterial3 = !_useMaterial3),
                                width: 80,
                                background: Color.Parse("#FFEAF6F7")),
                        ]),
                    new Expanded(
                        child: new SingleChildScrollView(
                            child: new Column(
                                crossAxisAlignment: CrossAxisAlignment.Stretch,
                                spacing: 14,
                                children:
                                [
                                    BuildShadeStrip(name, swatch),
                                    BuildThemeProbe(pageTheme),
                                ]))),
                ]));
    }

    private Widget BuildShadeStrip(string name, MaterialColor swatch)
    {
        List<Widget> tiles = [];
        foreach (int shade in Shades)
        {
            Color color = swatch[shade]!.Value;
            tiles.Add(new Expanded(
                child: new Container(
                    height: 56,
                    color: color,
                    alignment: Alignment.Center,
                    child: new Text(
                        shade.ToString(),
                        fontSize: 11,
                        color: ThemeData.EstimateBrightnessForColor(color) == Brightness.Dark
                            ? MaterialColors.White
                            : MaterialColors.Black))));
        }

        return new Container(
            color: Color.Parse("#FFF7F9FC"),
            padding: new Thickness(12),
            child: new Column(
                crossAxisAlignment: CrossAxisAlignment.Stretch,
                spacing: 8,
                children:
                [
                    new Text($"Colors.{name} shades", fontSize: 14, color: Colors.Black),
                    new Row(children: tiles),
                ]));
    }

    private Widget BuildThemeProbe(ThemeData theme)
    {
        return new Container(
            color: Color.Parse("#FFF7F9FC"),
            padding: new Thickness(12),
            child: new Column(
                crossAxisAlignment: CrossAxisAlignment.Stretch,
                spacing: 8,
                children:
                [
                    new Text("Theme colors", fontSize: 14, color: Colors.Black),
                    BuildSwatchRow("colorScheme.primary", theme.ColorScheme.Primary),
                    BuildSwatchRow("colorScheme.secondary", theme.ColorScheme.Secondary),
                    BuildSwatchRow("primaryColor", theme.PrimaryColor),
                    BuildSwatchRow("primaryColorLight", theme.PrimaryColorLight),
                    BuildSwatchRow("primaryColorDark", theme.PrimaryColorDark),
                    BuildSwatchRow("canvasColor", theme.CanvasColor),
                    BuildSwatchRow("cardColor", theme.CardColor),
                ]));
    }

    private Widget BuildSwatchRow(string label, Color color)
    {
        return new Row(
            spacing: 10,
            children:
            [
                new Container(width: 28, height: 20, color: color),
                new Expanded(child: new Text(label, fontSize: 12, color: Colors.Black)),
                new Text(
                    $"#{color.ToUInt32():X8}",
                    fontSize: 12,
                    color: Color.Parse("#FF607D8B")),
            ]);
    }

    private Widget BuildControlButton(
        string label,
        Action onTap,
        double width,
        Color background)
    {
        return new SizedBox(
            width: width,
            child: new TextButton(
                onPressed: onTap,
                child: new Text(label, fontSize: 12),
                style: TextButton.StyleFrom(
                    foregroundColor: Colors.Black,
                    backgroundColor: background,
                    padding: new Thickness(10, 8),
                    minimumSize: new Size(64, 36),
                    shape: new RoundedRectangleBorder(
                        borderRadius: BorderRadius.Circular(8)))));
    }
}
